### PR TITLE
Implement all compression types

### PR DESCRIPTION
**Pull request by astrofrog**
This is an experiment with adding a C extension to interface with the cfitsio PLIO_1 compression/decompression functions. Decompression seems to work (the test passes), compression does not work properly yet (the test fails). A few notes:

* When decompressing, we need to tell the C code how many elements we expect in the output so as to allocate the buffer, so we need to figure out what the API is for passing tile size down to the functions we are adding here.
* When decompressing, the C code always return 32-bit values that we should then cast to whatever is needed hence the if statement in the test. Presumably we should be returning a buffer from the decompression functions with the itemsize set correctly to reflect the 32-bit size of each item here.
* When compressing, we need to somehow guess how big to make the compressed data buffer, and I'm not sure what is the safest value to use here. Is it safe for example to assume the compressed bytes will always be shorter than the original data?

### CODE REVIEW

File: astropy/io/fits/setup_package.py [hunk 1]
```diff
@@ -68,5 +68,14 @@ def _get_compression_extension():
     return Extension('astropy.io.fits.compression', **cfg)
 
 
+def _get_simple_compression_extension():
+    return Extension('astropy.io.fits.simple_compression',
```
Comment (Cadair) [inline astropy/io/fits/setup_package.py:72]: Would it be better to put this and the python API in the same subpackage?

Comment (astrofrog) [reply astropy/io/fits/setup_package.py:72]: Yes

**Comment by astrofrog** [comment]
@Cadair - I've now re-organized into a sub-package and also added the C file which I had forgotten to add before 🤦

**Comment by astrofrog** [comment]
The PLIO test is still failing for the compression part (but the decompression is working).

**Comment by astrofrog** [comment]
The compression is almost working but a couple of values differ:

```
E        x: array([[ 0,  0,  0,  3,  4,  5,  6,  7,  8],
E              [ 0,  0,  0, 12, 13, 14, 15, 16, 17],
E              [ 0,  0,  0, 21, 22, 23, 24, 25, 26],...
E        y: array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8],
E              [ 9, 10, 11, 12, 13, 14, 15, 16, 17],
E              [18, 19, 20, 21, 22, 23, 24, 25, 26],...
```

I can try and investigate one evening soon but in the mean time feel free to merge if you need to push ahead with other parts (or know how to fix this).

**Comment by astrofrog** [comment]
@Cadair - I've now rebased and updated this, and I've also added support for RICE_1. The code/test is still quite messy at the moment as I'm prototyping and understanding a bit what info we need to pass down to the codecs. I can try and add HCOMPRESS and then tidy up after that at some point in the next few days.

**Comment by astrofrog** [comment]
Ok so I have ``test_basic`` passing for all algorithms! However test_basic is now a monster and includes a lot of logic that should live in the codecs, so will need to tidy that up (which will depend on the discussion in https://github.com/aperiosoftware/astropy/pull/17).

The C code also has a lot of unused variables etc so I'll need to clean that up.

**Comment by astrofrog** [comment]
I've now cleaned up the C code and fixed segfaults I was seeing with clang. I still see issues if I make some of the buffers too small so for now some of the buffer memory allocations are too large (and I've noted this in the C code). I think this can be investigated later. We can maybe consider merging this is tests pass for you too locally @Cadair, and then improve the API in a follow-up?

File: cextern/cfitsio/lib/ricecomp.h [hunk 2]
```diff
@@ -0,0 +1,6 @@
+int fits_rcomp_byte(signed char a[], int nx, unsigned char *c, int clen, int nblock);
+int fits_rcomp_short(short a[], int nx, unsigned char *c, int clen, int nblock);
+int fits_rcomp(int a[], int nx, unsigned char *c, int clen, int nblock);
+int fits_rdecomp_byte(unsigned char *c,	int clen, unsigned char array[], int nx, int nblock);
+int fits_rdecomp_short(unsigned char *c, int clen, unsigned short array[], int nx, int nblock);
+int fits_rdecomp(unsigned char *c, int clen, unsigned int array[],	int nx, int nblock);
```
Comment (Cadair) [inline cextern/cfitsio/lib/ricecomp.h:6]:
```suggestion
int fits_rdecomp(unsigned char *c, int clen, unsigned int array[], int nx, int nblock);
```

**Review by Cadair — APPROVED** [review]
That C code is hard work to understand lol.

**Comment by Cadair** [comment]
Looks like the CI has thrown some legitimate fails? https://github.com/aperiosoftware/astropy/actions/runs/3429073177/jobs/5714198212#step:6:741

File: astropy/io/fits/tiled_compression/src/compression.c [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+// Some of the cfitsio compression files use ffpmsg
+// so we provide a dummy function to replace this.
+void ffpmsg(const char *err_message) {}
+
+// Compatibility code because we pick up fitsio2.h from cextern. Can
+// remove once we remove cextern
+#ifdef _REENTRANT
+pthread_mutex_t Fitsio_Lock;
+int Fitsio_Pthread_Status = 0;
+#endif
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/src/compression.c:24]: Oh wow :grinning:

**Review by Cadair — APPROVED** [review]

**Pull request merged**